### PR TITLE
Intrinsify Interlocked.CompareExchange and Interlocked.Exchange for null

### LINKED ISSUE #69000: Perf: Interlocked.Exchange should special-case nullptr values
If you look at [the callers of `Interlocked.Exchange<T>(ref T, T)`](https://source.dot.net/System.Private.CoreLib/R/19d00939423c949f.html), you'll notice that a large number of them pass _nullptr_ as for the second parameter.

The method currently always results in a call to the native runtime, where the thread is notified of object assignment and the card table is updated if needed. However, if the value is known at JIT time to be _nullptr_, both of these steps can be skipped. This would allow the JIT to emit an `xchg` instruction directly into the codegen, avoiding the call back into the runtime. There's a similar optimization in place today when assigning a known-_nullptr_ value to a field; the JIT emits the `mov` instruction directly without bouncing through the internal "volatile write \& update card table" helper.

### PR DESCRIPTION
**Pull request by EgorBo**
Closes https://github.com/dotnet/runtime/issues/69000

```csharp
object Foo(ref object a)
    => Interlocked.Exchange(ref a, null);

object Foo(ref object a, object b)
    => Interlocked.CompareExchange(ref a, null, b);
}
```
Was:
```asm
; Assembly listing for method Prog:Foo(byref):System.Object:this
       488BCA               mov      rcx, rdx
       33D2                 xor      rdx, rdx
       E93616AC5F           jmp      Exchange(byref,System.Object):System.Object
; Total bytes of code 10


; Assembly listing for method Prog:Foo(byref,System.Object):System.Object:this
       488BCA               mov      rcx, rdx
       498BD0               mov      rdx, r8
       4533C0               xor      r8, r8
       E9A216AC5F           jmp      CompareExchange(byref,System.Object,System.Object):System.Object
; Total bytes of code 14
```
Now:
```asm
; Method Prog:Foo(byref):System.Object:this
       33C0                 xor      rax, rax
       488702               xchg     gword ptr [rdx], rax
       C3                   ret      
; Total bytes of code: 6


; Method Prog:Foo(byref,System.Object):System.Object:this
       33C9                 xor      rcx, rcx
       498BC0               mov      rax, r8
       F0                   lock     
       480FB10A             cmpxchg  gword ptr [rdx], rcx
       C3                   ret         
; Total bytes of code: 11
```

### CODE REVIEW

**Comment by jkotas** [comment]
> Interlocked.CompareExchange(ref a, b, null);

This needs write barrier for `b`. I do not think you can optimize it like this.

You can optimize `Interlocked.CompareExchange(ref a, null, b)`. It is much less common though.

File: src/coreclr/System.Private.CoreLib/src/System/Threading/Interlocked.CoreCLR.cs [hunk 1]
```diff
@@ -81,6 +81,7 @@ public static partial class Interlocked
         /// <param name="value">The value to which the <paramref name="location1"/> parameter is set.</param>
         /// <returns>The original value of <paramref name="location1"/>.</returns>
         /// <exception cref="NullReferenceException">The address of location1 is a null pointer.</exception>
+        [Intrinsic]
```
Comment (jkotas) [inline src/coreclr/System.Private.CoreLib/src/System/Threading/Interlocked.CoreCLR.cs:84]: It should be marked in nativeaot copy as well.

Comment (EgorBo) [reply src/coreclr/System.Private.CoreLib/src/System/Threading/Interlocked.CoreCLR.cs:84]: For some reason they're already marked as so there 🙂

**Comment by EgorBo** [comment]
> > Interlocked.CompareExchange(ref a, b, null);
> 
> This needs write barrier for `b`. I do not think you can optimize it like this.
> 
> You can optimize `Interlocked.CompareExchange(ref a, null, b)`. It is much less common though.

Thanks! Fixed, grep: https://grep.app/search?q=CompareExchange%5C%28ref%20%5Cw%2B%2C%20null%2C%20&regexp=true

**Comment by EgorBo** [comment]
@dotnet/jit-contrib PTAL, simple change

**Review by kunalspathak — APPROVED** [review]
LGTM. Thanks!

**Comment by EgorBo** [comment]
Wow, quite a few improvements: https://github.com/dotnet/perf-autofiling-issues/issues/10542 (e.g. Regex IsMatch as was expected by @GrabYourPitchforks in https://github.com/dotnet/runtime/issues/69000#issuecomment-1120129699)

**Comment by drieseng** [comment]
@EgorBo, is it expected that it **IsMatch_Multithreading** went from **56.73 ms** to **0 ns**?

**Comment by stephentoub** [comment]
> is it expected that IsMatch_Multithreading went from 56.73 ms to 0 ns?

No.  That's got to be a test harness / reporting issue, or else some kind of catastrophic failure :)

> Wow, quite a few improvements

`Interlocked.Exchange(ref location, null)` is reasonably common in multi-threaded caches / object pools, used to reserve some pooled object from a shared cache.  I'd expect to see wins in such uses (which many of the cited improvements seem to be).

**Pull request merged**